include/auto_tuner/gemm_tuner.hpp

```diff
@@ -0,0 +1,301 @@
+
+    auto accA = make_matrix_view(ex, m_a_gpu, m, k, lda, Access::ColMajor());
+    auto accB = make_matrix_view(ex, m_b_gpu, k, n, ldb, Access::ColMajor());
+    auto accC = make_matrix_view(ex, m_c_gpu, m, n, ldc, Access::ColMajor());
+    auto gemm = Gemm(accA, accB, accC, alpha, beta);
+    run_test(r, 2.0 * m * n * k, [&] {
+      auto event = ex.gemm_executor(gemm);
+      ex.wait(event);
+    });
+  }
+  std::cout << "err  = " << relative_diff(refC, dataC) << std::endl;
+}
+
+template <typename T, typename Container, typename Executor>
+void test_syclblas(int r, char transA, char transB, int m, int n, int k,
+                   T alpha, const Container &dataA, int lda,
+                   const Container &dataB, int ldb, T beta, Container dataC,
+                   int ldc, const Container &refC, Executor &ex) {
+  using etype = typename Container::value_type;
+  std::cout << "\n=== Testing SYCL-BLAS gemm ===" << std::endl;
+  {
+    auto m_a_gpu = blas::helper::make_sycl_iterator_buffer<etype>(
+        const_cast<etype *>(dataA.data()), dataA.size());
+    auto m_b_gpu = blas::helper::make_sycl_iterator_buffer<etype>(
+        const_cast<etype *>(dataB.data()), dataB.size());
+    auto m_c_gpu = blas::helper::make_sycl_iterator_buffer<etype>(
+        const_cast<etype *>(dataC.data()), dataC.size());
+    run_test(r, 2.0 * m * n * k, [&] {
+      auto event = _gemm(ex, transA, transB, m, n, k, alpha, m_a_gpu, lda,
+                         m_b_gpu, ldb, beta, m_c_gpu, ldc);
+      ex.wait(event);
+    });
+  }
+  std::cout << "err  = " << relative_diff(refC, dataC) << std::endl;
+}
+
+template <bool TransA, bool TransB, typename E>
+void run_gemm_tests(int seed, int m, int k, int n, int rep) {
+  std::cout << std::scientific;
+
+  std::mt19937 rnd(seed);
+
+  auto dataA = gen_matrix<E>(TransA ? k : m, TransA ? m : k, -1, 1, rnd);
+  auto dataB = gen_matrix<E>(TransB ? n : k, TransB ? k : n, -1, 1, rnd);
+  auto origC = gen_matrix<E>(m, n, -1, 1, rnd);
+  auto refC = origC;
+
+  const char *ta_str = TransA ? "T" : "N";
+  const char *tb_str = TransB ? "T" : "N";
+
+  const int lda = TransA ? k : m;
+  const int ldb = TransB ? n : k;
+  const int ldc = m;
+
+  std::cout << "\n=== Testing system CPU implementation ===" << std::endl;
+  run_test(rep, 2.0 * m * n * k, [&] {
+    gemm(ta_str, tb_str, m, n, k, E(1), dataA.data(), lda, dataB.data(), ldb,
+         E(1), refC.data(), m);
+  });
+
+  cl::sycl::queue q([=](cl::sycl::exception_list eL) {
+    try {
+      for (auto &e : eL) {
+        std::rethrow_exception(e);
+      }
+    } catch (cl::sycl::exception &e) {
+      std::cout << " E " << e.what() << std::endl;
+    } catch (...) {
+      std::cout << " An exception " << std::endl;
+    }
+  });
+  std::cout << "\nDevice: "
+            << q.get_device().get_info<cl::sycl::info::device::name>()
+            << std::endl;
+
+  Executor<SYCL> ex(q);
+
+#define ARG m, n, k, E(1), dataA, lda, dataB, ldb, E(1), origC, ldc, refC, ex
```

Comment:
This macro is too ugly, can you replace it by some function? You can also set this as a struct that you pass by value to all the test functions before. This macro is too insecure.